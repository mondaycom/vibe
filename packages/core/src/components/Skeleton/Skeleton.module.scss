@import "SkeletonVariables";

@mixin shine-animation() {
  overflow: hidden;
  animation-duration: 0.8s;
  animation-direction: alternate;
  animation-iteration-count: infinite;
  animation-name: shine;
  animation-timing-function: steps(10, end);

  @keyframes shine {
    0% {
      opacity: 0.4;
    }
    100% {
      opacity: 1;
    }
  }
}

@mixin skeleton($radius: var(--border-radius-small)) {
  background: var(--ui-background-color);
  border-radius: $radius;
  @include shine-animation;
}

.skeleton {
  display: flex;
}

.rectangle {
  height: $rectangle-default-size;
  @include skeleton;
}

.circle {
  height: $circle-default-size;
  @include skeleton($circle-radius);
}

.text {
  @include skeleton;
}

.textH1 {
  height: $h1-empty-state-height;
}

.textH2 {
  height: $h2-empty-state-height;
}

.textH3 {
  height: $h3-empty-state-height;
}

.textH4 {
  height: $h4-empty-state-height;
}

.textH5 {
  height: $h5-empty-state-height;
}

.textH6 {
  height: $h6-empty-state-height;
}

.textSmall,
.textCustom {
  height: $small-empty-state-height;
}

.fullWidth {
  width: 100%;
}

.skeleton:not(.fullWidth) {
  .text {
    width: 162px;
  }
  .circle {
    width: $circle-default-size;
  }
  .rectangle {
    width: $rectangle-default-size;
  }
}
