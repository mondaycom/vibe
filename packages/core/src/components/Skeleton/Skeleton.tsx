import { camelCase } from "lodash-es";
import { ComponentDefaultTestId, getTestId } from "../../tests/test-ids-utils";
import cx from "classnames";
import React, { FC } from "react";
import {
  SKELETON_SIZES,
  SkeletonType as SkeletonTypeEnum,
  SKELETON_CUSTOM_SIZE,
  TextSkeletonSize as TextSkeletonSizeEnum
} from "./SkeletonConstants";

import { SkeletonType, SkeletonSizeType } from "./Skelton.types";
import { getStyle } from "../../helpers/typesciptCssModulesHelper";
import { withStaticProps, VibeComponentProps } from "../../types";
import styles from "./Skeleton.module.scss";

export interface SkeletonProps extends VibeComponentProps {
  /**
   * The type of skeleton.
   */
  type?: SkeletonType;
  /**
   * The predefined size of the skeleton.
   */
  size?: SkeletonSizeType;
  /**
   * The width of the skeleton in pixels.
   */
  width?: number;
  /**
   * The height of the skeleton in pixels.
   */
  height?: number;
  /**
   * Class name applied to the wrapper element.
   */
  wrapperClassName?: string;
  /**
   * If true, the skeleton will take up the full width of its container.
   */
  fullWidth?: boolean;
}

const Skeleton: FC<SkeletonProps> & {
  types?: typeof SkeletonTypeEnum;
  sizes?: typeof SKELETON_SIZES;
} = ({
  type = "rectangle",
  size = "custom",
  className,
  wrapperClassName,
  width,
  height,
  fullWidth = false,
  id,
  "data-testid": dataTestId
}) => {
  const skeletonType = (Object.values(SkeletonTypeEnum) as string[]).includes(type) ? type : "rectangle";

  // Skeleton has sizes only for text type, other types support only custom size
  const skeletonSize = (Object.values(TextSkeletonSizeEnum) as string[]).includes(size) ? size : SKELETON_CUSTOM_SIZE;

  return (
    <div
      id={id}
      className={cx(styles.skeleton, wrapperClassName, { [styles.fullWidth]: fullWidth })}
      data-testid={dataTestId || getTestId(ComponentDefaultTestId.SKELETON, id)}
    >
      <div
        className={cx(styles[skeletonType], getStyle(styles, camelCase(skeletonType + "-" + skeletonSize)), className, {
          [styles.fullWidth]: fullWidth
        })}
        style={{ width, height }}
      />
    </div>
  );
};

export default withStaticProps(Skeleton, {
  types: SkeletonTypeEnum,
  sizes: SKELETON_SIZES
});
