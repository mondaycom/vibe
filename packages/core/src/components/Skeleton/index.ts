export { default as Skeleton, type SkeletonProps } from "./Skeleton";

export * from "./Skelton.types";
