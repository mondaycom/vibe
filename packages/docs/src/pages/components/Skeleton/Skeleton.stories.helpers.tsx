import React from "react";
import { StorybookLink, Tip } from "vibe-storybook-components";

export const TipLoader = () => (
  <Tip>
    Using loader after the pages are fully loaded consider using the{" "}
    <StorybookLink page="Components/Loader" size={StorybookLink.sizes.SMALL}>
      loader component.
    </StorybookLink>
  </Tip>
);
