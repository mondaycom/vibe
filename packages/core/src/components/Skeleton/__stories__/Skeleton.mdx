import Skeleton from "../Skeleton";
import { Meta } from "@storybook/blocks";
import {
  COUNTER,
  LINEAR_PROGRESS_BAR,
  SPINNER
} from "../../../storybook/components/related-components/component-description-map";
import { TipLoader } from "./Skeleton.stories.helpers";
import * as SkeletonStories from "./Skeleton.stories";
import { Flex } from "../../Flex";

<Meta of={SkeletonStories} />

# Skeleton

- [Overview](#overview)
- [Props](#props)
- [Usage](#usage)
- [Variants](#variants)
- [Do’s and don’ts](#dos-and-donts)
- [Use cases and examples](#use-cases-and-examples)
- [Related components](#related-components)
- [Feedback](#feedback)

## Overview

Skeleton loading componet used to indicate content and ui loading that will appear after its loaded. It helps to decrease perceived duration time

<Canvas of={SkeletonStories.Overview} />

### Import

```js
import { Skeleton } from "@vibe/core";
```

## Props

<PropsTable />

## Usage

<UsageGuidelines
  guidelines={[
    "Mimic the user interface that is going to be loaded as much as possible",
    "Animation on the skeleton is a repeating color pulse should move from left to right to indicate the lading state"
  ]}
/>

<TipLoader />

## Variants

### Shapes

Use shaes to mimic Avatars, images, buttons etc...

<Canvas of={SkeletonStories.Shapes} />

### Text

Presents a classic menu or equivalent picker

<Canvas of={SkeletonStories.TextSkeleton} />

## Do’s and Don’ts

<ComponentRules
  rules={[
    {
      positive: {
        component: (
          <Flex gap="small" direction="column" align="start">
            <Flex gap="small">
              <Skeleton type="circle" />
              <Skeleton type="text" size="small" />
            </Flex>
            <Flex gap="small">
              <Skeleton type="circle" />
              <Skeleton type="text" size="small" />
            </Flex>
            <Flex gap="small">
              <Skeleton type="circle" />
              <Skeleton type="text" size="small" />
            </Flex>
          </Flex>
        ),
        description: "Show only representation of ui."
      },
      negative: {
        component: (
          <Flex gap="small" direction="column" align="start">
            <Flex gap="small">
              <Skeleton type="circle" />
              <div>
                <Skeleton type="text" size="small" />
                Small text
              </div>
            </Flex>
            <Flex gap="small">
              <Skeleton type="circle" />
              <div>
                <Skeleton type="text" size="small" />
                Small text
              </div>
            </Flex>
            <Flex gap="small">
              <Skeleton type="circle" />
              <div>
                <Skeleton type="text" size="small" />
                Small text
              </div>
            </Flex>
          </Flex>
        ),
        description: "Show all the user interface exactly."
      }
    }
  ]}
/>

## Use cases and examples

### Complex Example

<Canvas of={SkeletonStories.ComplexExample} />

### Update in the system

Use this menu to allow a user to either select one or more items from the list.

<Canvas of={SkeletonStories.UpdateInTheSystem} />

## Related components

<RelatedComponents componentsNames={[LINEAR_PROGRESS_BAR, COUNTER, SPINNER]} />
