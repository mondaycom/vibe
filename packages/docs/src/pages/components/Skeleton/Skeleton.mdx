import { Skeleton, Flex } from "@vibe/core";
import { Meta } from "@storybook/blocks";
import { TipLoader } from "./Skeleton.stories.helpers";
import * as SkeletonStories from "./Skeleton.stories";

<Meta of={SkeletonStories} />

# Skeleton

Skeleton loading component used to indicate content and ui loading that will appear after its loaded. It helps to decrease perceived duration time

<Canvas of={SkeletonStories.Overview} />

### Import

```js
import { Skeleton } from "@vibe/core";
```

## Props

<PropsTable />

## Usage

<UsageGuidelines
  guidelines={[
    "Mimic the user interface that is going to be loaded as much as possible",
    "Animation on the skeleton is a repeating color pulse should move from left to right to indicate the lading state"
  ]}
/>

<TipLoader />

## Variants

### Shapes

Use shapes to mimic Avatars, images, buttons etc...

<Canvas of={SkeletonStories.Shapes} />

### Text

Presents a classic menu or equivalent picker

<Canvas of={SkeletonStories.TextSkeleton} />

## Do’s and Don’ts

<ComponentRules
  rules={[
    {
      positive: {
        component: (
          <Flex gap="small" direction="column" align="start">
            <Flex gap="small">
              <Skeleton type="circle" />
              <Skeleton type="text" size="small" />
            </Flex>
            <Flex gap="small">
              <Skeleton type="circle" />
              <Skeleton type="text" size="small" />
            </Flex>
            <Flex gap="small">
              <Skeleton type="circle" />
              <Skeleton type="text" size="small" />
            </Flex>
          </Flex>
        ),
        description: "Show only representation of ui."
      },
      negative: {
        component: (
          <Flex gap="small" direction="column" align="start">
            <Flex gap="small">
              <Skeleton type="circle" />
              <div>
                <Skeleton type="text" size="small" />
                Small text
              </div>
            </Flex>
            <Flex gap="small">
              <Skeleton type="circle" />
              <div>
                <Skeleton type="text" size="small" />
                Small text
              </div>
            </Flex>
            <Flex gap="small">
              <Skeleton type="circle" />
              <div>
                <Skeleton type="text" size="small" />
                Small text
              </div>
            </Flex>
          </Flex>
        ),
        description: "Show all the user interface exactly."
      }
    }
  ]}
/>

## Use cases and examples

### Complex Example

<Canvas of={SkeletonStories.ComplexExample} />

### Update in the system

Use this menu to allow a user to either select one or more items from the list.

<Canvas of={SkeletonStories.UpdateInTheSystem} />

## Related components

<RelatedComponents componentsNames={["LinearProgressBar", "Counter", "Spinner"]} />
