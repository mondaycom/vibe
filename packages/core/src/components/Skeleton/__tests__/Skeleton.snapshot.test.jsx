import { describe, it, expect } from "vitest";
import React from "react";
import renderer from "react-test-renderer";
import Skeleton from "../Skeleton";

describe("Skeleton renders correctly", () => {
  it("renders correctly with empty props", () => {
    const tree = renderer.create(<Skeleton />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe("renders correctly when rectangle", () => {
    it("with empty props", () => {
      const tree = renderer.create(<Skeleton type={Skeleton.types.RECTANGLE} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with fixed height", () => {
      const tree = renderer.create(<Skeleton type={Skeleton.types.RECTANGLE} height={100} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with fixed width", () => {
      const tree = renderer.create(<Skeleton type={Skeleton.types.RECTANGLE} width={100} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with fixed height and width", () => {
      const tree = renderer.create(<Skeleton type={Skeleton.types.RECTANGLE} width={100} height={100} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("renders correctly when circle", () => {
    it("with empty props", () => {
      const tree = renderer.create(<Skeleton type="circle" />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with fixed height", () => {
      const tree = renderer.create(<Skeleton type="circle" height={100} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with fixed width", () => {
      const tree = renderer.create(<Skeleton type="circle" width={100} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with fixed height and width", () => {
      const tree = renderer.create(<Skeleton type="circle" width={100} height={100} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("renders correctly when text", () => {
    it("with empty props", () => {
      const tree = renderer.create(<Skeleton type={Skeleton.types.TEXT} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with fixed height", () => {
      const tree = renderer.create(<Skeleton type={Skeleton.types.TEXT} height={100} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with fixed width", () => {
      const tree = renderer.create(<Skeleton type={Skeleton.types.TEXT} width={100} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with fixed height and width", () => {
      const tree = renderer.create(<Skeleton type={Skeleton.types.TEXT} width={100} height={100} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with H1 size", () => {
      const tree = renderer.create(<Skeleton type={Skeleton.types.TEXT} size={Skeleton.types.H1} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with H2 size", () => {
      const tree = renderer.create(<Skeleton type={Skeleton.types.TEXT} size={Skeleton.types.H2} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with H3 size", () => {
      const tree = renderer.create(<Skeleton type={Skeleton.types.TEXT} size={Skeleton.types.H3} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with H4 size", () => {
      const tree = renderer.create(<Skeleton type={Skeleton.types.TEXT} size={Skeleton.types.H4} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with H5 size", () => {
      const tree = renderer.create(<Skeleton type={Skeleton.types.TEXT} size={Skeleton.types.H5} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with CUSTOM size", () => {
      const tree = renderer.create(<Skeleton type={Skeleton.types.TEXT} size={Skeleton.sizes.CUSTOM} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
