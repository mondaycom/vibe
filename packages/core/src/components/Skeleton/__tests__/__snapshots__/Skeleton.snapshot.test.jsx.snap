// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Skeleton renders correctly > renders correctly when circle > with empty props 1`] = `
<div
  className="skeleton"
  data-testid="skeleton"
>
  <div
    className="circle circleCustom"
    style={
      {
        "height": undefined,
        "width": undefined,
      }
    }
  />
</div>
`;

exports[`Skeleton renders correctly > renders correctly when circle > with fixed height 1`] = `
<div
  className="skeleton"
  data-testid="skeleton"
>
  <div
    className="circle circleCustom"
    style={
      {
        "height": 100,
        "width": undefined,
      }
    }
  />
</div>
`;

exports[`Skeleton renders correctly > renders correctly when circle > with fixed height and width 1`] = `
<div
  className="skeleton"
  data-testid="skeleton"
>
  <div
    className="circle circleCustom"
    style={
      {
        "height": 100,
        "width": 100,
      }
    }
  />
</div>
`;

exports[`Skeleton renders correctly > renders correctly when circle > with fixed width 1`] = `
<div
  className="skeleton"
  data-testid="skeleton"
>
  <div
    className="circle circleCustom"
    style={
      {
        "height": undefined,
        "width": 100,
      }
    }
  />
</div>
`;

exports[`Skeleton renders correctly > renders correctly when rectangle > with empty props 1`] = `
<div
  className="skeleton"
  data-testid="skeleton"
>
  <div
    className="rectangle rectangleCustom"
    style={
      {
        "height": undefined,
        "width": undefined,
      }
    }
  />
</div>
`;

exports[`Skeleton renders correctly > renders correctly when rectangle > with fixed height 1`] = `
<div
  className="skeleton"
  data-testid="skeleton"
>
  <div
    className="rectangle rectangleCustom"
    style={
      {
        "height": 100,
        "width": undefined,
      }
    }
  />
</div>
`;

exports[`Skeleton renders correctly > renders correctly when rectangle > with fixed height and width 1`] = `
<div
  className="skeleton"
  data-testid="skeleton"
>
  <div
    className="rectangle rectangleCustom"
    style={
      {
        "height": 100,
        "width": 100,
      }
    }
  />
</div>
`;

exports[`Skeleton renders correctly > renders correctly when rectangle > with fixed width 1`] = `
<div
  className="skeleton"
  data-testid="skeleton"
>
  <div
    className="rectangle rectangleCustom"
    style={
      {
        "height": undefined,
        "width": 100,
      }
    }
  />
</div>
`;

exports[`Skeleton renders correctly > renders correctly when text > with CUSTOM size 1`] = `
<div
  className="skeleton"
  data-testid="skeleton"
>
  <div
    className="text textCustom"
    style={
      {
        "height": undefined,
        "width": undefined,
      }
    }
  />
</div>
`;

exports[`Skeleton renders correctly > renders correctly when text > with H1 size 1`] = `
<div
  className="skeleton"
  data-testid="skeleton"
>
  <div
    className="text textCustom"
    style={
      {
        "height": undefined,
        "width": undefined,
      }
    }
  />
</div>
`;

exports[`Skeleton renders correctly > renders correctly when text > with H2 size 1`] = `
<div
  className="skeleton"
  data-testid="skeleton"
>
  <div
    className="text textCustom"
    style={
      {
        "height": undefined,
        "width": undefined,
      }
    }
  />
</div>
`;

exports[`Skeleton renders correctly > renders correctly when text > with H3 size 1`] = `
<div
  className="skeleton"
  data-testid="skeleton"
>
  <div
    className="text textCustom"
    style={
      {
        "height": undefined,
        "width": undefined,
      }
    }
  />
</div>
`;

exports[`Skeleton renders correctly > renders correctly when text > with H4 size 1`] = `
<div
  className="skeleton"
  data-testid="skeleton"
>
  <div
    className="text textCustom"
    style={
      {
        "height": undefined,
        "width": undefined,
      }
    }
  />
</div>
`;

exports[`Skeleton renders correctly > renders correctly when text > with H5 size 1`] = `
<div
  className="skeleton"
  data-testid="skeleton"
>
  <div
    className="text textCustom"
    style={
      {
        "height": undefined,
        "width": undefined,
      }
    }
  />
</div>
`;

exports[`Skeleton renders correctly > renders correctly when text > with empty props 1`] = `
<div
  className="skeleton"
  data-testid="skeleton"
>
  <div
    className="text textCustom"
    style={
      {
        "height": undefined,
        "width": undefined,
      }
    }
  />
</div>
`;

exports[`Skeleton renders correctly > renders correctly when text > with fixed height 1`] = `
<div
  className="skeleton"
  data-testid="skeleton"
>
  <div
    className="text textCustom"
    style={
      {
        "height": 100,
        "width": undefined,
      }
    }
  />
</div>
`;

exports[`Skeleton renders correctly > renders correctly when text > with fixed height and width 1`] = `
<div
  className="skeleton"
  data-testid="skeleton"
>
  <div
    className="text textCustom"
    style={
      {
        "height": 100,
        "width": 100,
      }
    }
  />
</div>
`;

exports[`Skeleton renders correctly > renders correctly when text > with fixed width 1`] = `
<div
  className="skeleton"
  data-testid="skeleton"
>
  <div
    className="text textCustom"
    style={
      {
        "height": undefined,
        "width": 100,
      }
    }
  />
</div>
`;

exports[`Skeleton renders correctly > renders correctly with empty props 1`] = `
<div
  className="skeleton"
  data-testid="skeleton"
>
  <div
    className="rectangle rectangleCustom"
    style={
      {
        "height": undefined,
        "width": undefined,
      }
    }
  />
</div>
`;
